stan/math/rev/core/deep_copy_vars.hpp

```diff
@@ -20,7 +20,7 @@ namespace math {
  *  Otherwise it will be moved.
  */
 template <typename Arith, typename = require_arithmetic_t<scalar_type_t<Arith>>>
-inline decltype(auto) deep_copy_vars(Arith&& arg) {
+inline Arith deep_copy_vars(Arith&& arg) {
```

Comment:
Was the bug that `deep_copy` wasn't handling expressions correctly? Or was it that the `reduce_sum` threaded signature didn't work with expressions?

Should there be a changes to Jenkins that turns on threading for the expression tests?

Reply:
Threaded expressions tests for reduce_sum/map_rect will be added (that needs to be done in the Stan repo). I will add that later today. But that would not catch one of the two issues. 


For the second one need to add a minimal version of the model to the test models.

Reply:
I'm asking cause it looks like `deep_copy_vars` was just returning the wrong type or something.

If it's that, then we just need to add something like this to the `deep_copy_var` unit tests:

```
EXPECT_TRUE((std::is_same<decltype(deep_copy_vars(arith_expr)), plain_type_t<decltype(arith_expr)>>));
```

Reply:
Well that seems wrong to me now cause it's assuming that deep_copy_vars evals the expression -- anyway, something like this? It looks like a problem with `deep_copy_vars` and I'm not sure what exactly `decltype(auto)` gets wrong there (though I don't doubt it gets something wrong).

Reply:
>Was the bug that deep_copy wasn't handling expressions correctly?

No. The issue was `reduce_sum` was assuming `deep_copy` does not return rvalue references. Since returning rvalue references can in this case cause some other issues, I just fixed it here.

>Well that seems wrong to me now cause it's assuming that deep_copy_vars evals the expression

Nope, there is no `eval` or `plain_type_t` in `deep_copy_vars`, so it does not eval anything.

Reply:
This makes sense.